src/parquet/bloom.h

```diff
@@ -0,0 +1,191 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PARQUET_BLOOM_H
+#define PARQUET_BLOOM_H
+
+#include <set>
+#include <cstdint>
+
+#include "parquet/types.h"
+#include "parquet/util/memory.h"
+
+namespace parquet {
+class OutputStream;
+
+// Bloom Filter is a compact structure to indicate whether an item is not in set or
+// probably in set. Bloom class is underlying class of Bloom Filter which stores a
+// bit set represents elements set, hash strategy and bloom filter algorithm.
+
+// Bloom Filter algorithm is implemented using block Bloom filters from Putze et al.'s
+// "Cache-,Hash- and Space-Efficient Bloom Filters". The basic idea is to hash the
+// item to a tiny Bloom Filter which size fit a single cache line or smaller. This
+// implementation sets 8 bits in each tiny Bloom Filter. Tiny bloom filter are 32
+// bytes to take advantage of 32-bytes SIMD instruction.
+
+class Bloom final {
+ public:
+  // Hash strategy available for bloom filter.
+  enum HashStrategy {
+    MURMUR3_X64_128
+  };
+
+  // Bloom filter algorithm.
+  enum Algorithm {
+    BLOCK
+  };
+
+  /**
+   * Default false positive probability value use to calculate optimal number of bits
+   * used by bloom filter.
+   */
+  static constexpr double DEFAULT_FPP = 0.01;
+
+  // Bloom filter data header, including number of bytes, hash strategy and algorithm.
+  static constexpr int HEADER_SIZE = 12;
+
+  // Bytes in a tiny bloom filter block.
+  static constexpr int BYTES_PER_FILTER_BLOCK = 32;
+
+  // Default seed for hash function, to keep cross compatibility the seed is same as
+  // value defined in parquet-mr.
+  static constexpr int DEFAULT_SEED = 104729;
+
+  // Default maximum bloom filter size (need to discuss)
```

Comment:
Originally I think this would need to discuss with parquet community, maybe we can set to 1/8 or parquet.block.size in case of one column of long type with its distinct values ratio equal to 1 and FPP set to 0.01. What do you think?

Reply:
What would it mean to set this to 1/8? I don't understand. Doesn't this need to be an integer?

I would suggest starting the conversation on dev@ if you think it needs to be discussed by the community before committing.

Reply:
Yes, I will start thread on this.